server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java

```diff
@@ -603,15 +613,19 @@ public BigArrays bigArrays() {  // TODO this is only used in agg land, maybe rem
         return bigArrays;
     }
 
-    private static Map<String, MappedFieldType> parseRuntimeMappings(
-        Map<String, Object> runtimeMappings,
-        MapperService mapperService
-    ) {
+    private static Map<String, MappedFieldType> parseRuntimeMappings(Map<String, Object> runtimeMappings, MapperService mapperService) {
         Map<String, MappedFieldType> runtimeFieldTypes = new HashMap<>();
         if (runtimeMappings.isEmpty() == false) {
             RuntimeFieldType.parseRuntimeFields(new HashMap<>(runtimeMappings), mapperService.parserContext(),
                 runtimeFieldType -> runtimeFieldTypes.put(runtimeFieldType.name(), runtimeFieldType));
         }
         return Collections.unmodifiableMap(runtimeFieldTypes);
     }
+
+    /**
+     * Cache key for current mapping.
+     */
+    public MappingLookup.CacheKey mappingCacheKey() {
+        return mappingLookup.cacheKey();
```

Comment:
maybe I only need to get used to this approach, but I wonder if it is necessary to carry around this additional cache key. I guess it is no option to make it become part of the existing cacheKey that the request exposes or something along these lines?

Reply:
I have to carry it around somewhere for us be able to assert on it. Originally I was using a `Writeable` key and inserting it into the request's cache key. When I was just writing a long it was fine but writing the whole mapping was quite big. And it was incomplete because we couldn't possibly handle analyzer reloading. Using the dummy object is nicer because it is smaller and can handle the analyzer reloading.

But, yeah, there is a fair bit of plumbing to get it to a better spot. What were you thinking?

Reply:
I will defer to @jtibshirani on this. I don't have any specific idea, just noticing that while this approach has advantages, it does require some additional plumbing and complexity. Maybe it's the right compromise though. I will drop another question: do we need to worry about hash collisions?

Reply:
> I will drop another question: do we need to worry about hash collisions?

When we last talked with @jpountz he said something along the lines of "I don't want to ever has to double check that we're not getting a hash collision when helping someone debug something." The resonated with me. Hash collisions aren't likely at all but it gives us a nice piece of mind at a fairly low cost. We'd have to plumb *something* through anyway, I think. Going this way does require all those changes to the `IndicesRequestCache` but they weren't too bad. Probably worth the peace of mind.

Reply:
I liked how this matched our approach for incorporating `DirectoryReader` into the cache key. We use a dedicated object reference, instead of keying on the actual content of `MappingLookup`. Specifically, we're not keying on a hash, so we shouldn't need to handle hash collisions ourself.

I suggested this approach as an option, but not an expert in this caching code and would be happy if you wanted to check with someone more familiar!